### ISSUE TITLE
BUG: diff gives incorrect results for datetime column written with pyarrow 12.0.1 installed and read with pyarrow 13.0.0 installed

### ISSUE THREAD

**Issue by selasley** [pandas-dev/pandas#54777]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
# in a virtualenv with pyarrow 12.0.1 installed write a dataframe with a datetime column to
# a parquet file with to_parquet()
import pandas as pd
df = pd.DataFrame(index=range(4), data={'A': pd.date_range('2021-01-01', '2021-01-04', 4)})
df.to_parquet('/tmp/pyarrow12.pq')
df.dtypes
df.diff()
df['A'].diff()


# in a virtualenv with pyarrow 13.0.0 installed read the file with read_parquet()
# and diff() the dataframe
import pandas as pd
df = pd.read_parquet('/tmp/pyarrow12.pq')
df.dtypes 
df.diff()
df['A'].diff()
```

### Issue Description

In the pyarrow 12 virtualenv df.dtypes is datetime64[ns].  df.diff() and df['A'].diff both return 1 days

In the pyarrow 13 virtualenv df.dtypes is datetime64[us].  df.diff() returns 1 days 00:00:0 but df['A'].diff() returns 0 days 00:01:26.400000

### Expected Behavior

Expect df.diff() and df['A'].diff() to return 1 days regardless of which version of pyarrow was installed when the file was written and read

> 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.11.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.6.0
Version          : Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:53 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6020
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.3
numpy            : 1.25.2
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 68.1.2
pip              : 23.2.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.11.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.6.0
Version          : Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:53 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6020
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.3
numpy            : 1.25.2
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 68.1.2
pip              : 23.2.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 13.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

**Comment by selasley** [comment]
With pandas 2.1.0rc0 df.diff() and df['A'].diff() both return 1 days 00:00:00

**Comment by jorisvandenbossche** [comment]
@selasley thanks for the report. 

This is because it seems we had a bug in pandas 2.0, which is fixed in the upcoming 2.1:

```
In [2]: pd.__version__
Out[2]: '2.0.3'

In [3]: pd.Series(pd.date_range("2012-01-01", freq="D", periods=3).as_unit("ns"))
Out[3]: 
0   2012-01-01
1   2012-01-02
2   2012-01-03
dtype: datetime64[ns]

In [4]: pd.Series(pd.date_range("2012-01-01", freq="D", periods=3).as_unit("ns")).diff()
Out[4]: 
0      NaT
1   1 days
2   1 days
dtype: timedelta64[ns]

In [5]: pd.Series(pd.date_range("2012-01-01", freq="D", periods=3).as_unit("ms")).diff()
Out[5]: 
0                      NaT
1   0 days 00:00:00.086400
2   0 days 00:00:00.086400
dtype: timedelta64[ns]
```

So when having a datetime64 column with a non-nanosecond resolution, the `diff()` method gave wrong results. 

This is fixed on the latest dev version:

```
In [1]: pd.__version__
Out[1]: '2.1.0rc0+8.g1bfd932611'

In [2]: pd.Series(pd.date_range("2012-01-01", freq="D", periods=3).as_unit("ms")).diff()
Out[2]: 
0               NaT
1   1 days 00:00:00
2   1 days 00:00:00
dtype: timedelta64[ms]

In [3]: pd.Series(pd.date_range("2012-01-01", freq="D", periods=3).as_unit("ms"))
Out[3]: 
0   2012-01-01
1   2012-01-02
2   2012-01-03
dtype: datetime64[ms]
```

cc @jbrockmendel do you know if this was fixed intentionally? (eg in TimedeltaArray constructor) I don't directly see a whatsnew note about it or related issue. And if not we should ensure to add a test.

@selasley now, the reason that you see this bug depending on the pyarrow version is because starting with pyarrow 13.0.0, pyarrow will preserve the original resolution of the timestamp column. And when writing to a Parquet file with pyarrow < 13.0, the timestamps would not be stored with nanosecond resolution, and thus also come back as non-nanosecond when reading such file with pyarrow >= 13.0.0.

**Comment by jbrockmendel** [comment]
> do you know if this was fixed intentionally

I don't recall doing anything specific to diff recently
